### ISSUE TITLE
Add URL redirects

### ISSUE THREAD

**Issue by peterdesmet** [tdwg/dwc#72]
## Questions (to be reviewed)

- [x] For all redirects, anchors should continue to work, so http://rs.tdwg.org/dwc/terms/history/index.htm#acceptedScientificName-2009-07-06 should link to http://rs.tdwg.org/dwc/terms/history.html#acceptedScientificName-2009-07-06*
- [x] URLs for old directories (not ending in `.htm` or `.html`) should redirect to the correct location, so http://rs.tdwg.org/dwc/terms/simple/ should link to dwc.tdwg.org/simple/
- [x] I also noticed that http://rs.tdwg.org/dwc/terms/simple (no trailing slash) redirects to http://tdwg.github.io/dwc/terms/index.htm#simple, which is an artefact of it being seen as a term.

## Specific rules

- [x] http://rs.tdwg.org/dwc/ | dwc.tdwg.org/
- [x] http://rs.tdwg.org/dwc/index.htm | dwc.tdwg.org/
- [x] http://rs.tdwg.org/dwc/terms/ | dwc.tdwg.org/terms/
- [x] http://rs.tdwg.org/dwc/terms/index.htm | dwc.tdwg.org/terms/
- [x] http://rs.tdwg.org/dwc/terms/index.htm#basisOfRecord | dwc.tdwg.org/terms/#dwc:basisOfRecord (note: `dwc:`)
- [x] http://rs.tdwg.org/dwc/terms/basisOfRecord | dwc.tdwg.org/terms/#dwc:basisOfRecord
- [ ] http://rs.tdwg.org/dwc/dwcore/version/Family-2007-04-17 | Handled by @baskaufs script?  
- [x] http://rs.tdwg.org/dwc/terms/namespace/index.htm | dwc.tdwg.org/namespace/
- [x] http://rs.tdwg.org/dwc/terms/simple/index.htm | dwc.tdwg.org/simple/
- [x] http://rs.tdwg.org/dwc/terms/guides/text/index.htm | dwc.tdwg.org/text/
- [x] http://rs.tdwg.org/dwc/terms/guides/xml/index.htm | dwc.tdwg.org/xml/
- [x] http://rs.tdwg.org/dwc/terms/guides/rdf/index.htm | dwc.tdwg.org/rdf/
- [x] http://rs.tdwg.org/dwc/terms/history/index.htm | https://github.com/tdwg/dwc/blob/master/vocabulary/term_versions.csv
- [ ] http://rs.tdwg.org/dwc/terms/history/decisions/index.htm | dwc.tdwg.org/decisions/
- [ ] http://rs.tdwg.org/dwc/terms/history/decisions/#Decision-2009-12-07_1 | dwc.tdwg.org/decisions/#Decision-2009-12-07_1
- [x] http://rs.tdwg.org/dwc/terms/history/dwctoabcd/index.htm | https://github.com/tdwg/dwc/blob/master/vocabulary/term_versions.csv
- [x] http://rs.tdwg.org/dwc/terms/history/versions/index.htm | https://github.com/tdwg/dwc/blob/master/vocabulary/term_versions.csv

Also check if cross-domain issues remain for Font Awesome icons.

**Comment by peterdesmet** [comment]
The new Darwin Core website consists of a single page: the quick reference guide. In the future, it might consists of a homepage and more guides + examples, maintained by the community. With that in mind, do you agree that the quick reference guide will live at:

http://rs.tdwg.org/dwc/guides/index.html

With redirects in place from http://rs.tdwg.org/dwc/terms/index.htm. Or you prefer the current URL http://rs.tdwg.org/dwc/terms/index.html (with `html` rather than `htm` though) for the quick reference guide?

/cc @mdoering @tucotuco @timrobertson100

**Comment by baskaufs** [comment]
I think it would be good to do some careful planning about identifier construction before making a final decision about the URIs you listed in your "General Rules" section.  In particular, I'm concerned about your suggestions that place a number of pages below http://rs.tdwg.org/dwc/terms/ (mapping.html, decisions.html, etc.).  That's because all of the terms themselves are at that level, e.g. http://rs.tdwg.org/dwc/terms/continent .  So if we are serious about making every URI dereference with content negotiation, that's going to make writing the server script that handles the dereferencing more complicated.  When there is a pattern match for http://rs.tdwg.org/dwc/terms/*, it's going to have to do a bunch of ad hoc testing to find out if the resource is a web page before defaulting to a more general procedure for generating metadata about a term.  

I'm trying to think about general URI patterns that we might apply across TDWG vocabularies (existing and future).  I'll try to put those thought down soon.

**Comment by mdoering** [comment]
I agree with Steven. It would also be wise to not include the html suffix in links as its application specific and github resolves it for you. Similar index.html pages should not be part of the URL but just the directory name. So rather http://rs.tdwg.org/dwc/terms/ and http://rs.tdwg.org/dwc/guides/

**Comment by baskaufs** [comment]
What Markus suggested is a fairly widespread practice. See for example http://dublincore.org/documents/dcmi-terms/index.html and http://dublincore.org/documents/dcmi-terms/ and various W3C recommendations.

**Comment by baskaufs** [comment]
I have written a [detailed description of the repo that I created](https://github.com/baskaufs/tdwg-standards/blob/master/README.md) during the DwC hackathon.  On that page, I describe the levels of standards components as they are laid out in the Standards Documentation Specification.  (This was a topic that came up during the @tdwg/tag-team meeting that I remoted into last week.)  What is relevant to this issue is that the [document describes URI patterns](https://github.com/baskaufs/tdwg-standards/blob/master/README.md#patterns) that I used when I generated the metadata that's included in [the repo](https://github.com/baskaufs/tdwg-standards).  For current resources, I based those patterns on existing practice in Darwin Core and Audubon Core.  For [versions of resources, I established a convention](https://github.com/baskaufs/tdwg-standards/blob/master/README.md#patterns-versions) of putting "version/" in the URI one level above what is effectively the local name for the resource.  You can see what I mean by looking at examples in the document.  I developed that convention with an eye towards what would make it easiest for enabling a server to do URI pattern matching in a script or configuration file.  

So the question that's relevant to this issue is how document URIs might fit with the established patterns for vocabularies, term lists, and terms.  In terms of the hierarchy, in many cases, they fall directly below the standard itself.  In fact, some standards are composed of only a single document.  So it would seem to make sense that the URI pattern should involve an addition to the URI at a level below the standard URI.  However, for historical reasons, the "permanent URLs" of standards have been in the form of ```http://www.tdwg.org/standards/``` (that is, under the www.tdwg.org subdomain), while pretty much all other standards components are under the rs.tdwg.org subdomain.  We should not break the existing standards URIs, but I don't think its a good idea construct URIs for documents (that are at the hierarchical level directly below standard) by appending strings to the standards URI.  It would be best to just have them be in the rs.tdwg.org subdomain along with everything else.  

It wouldn't really fit with the URI patterns for vocabularys, term lists, and terms, but I suppose one could just stick documents under a subpath designator for the standard.  For example, although ```http://rs.tdwg.org/ac/``` represents the Audubon Core vocabulary rather than the Audubon Core standard (which is ```http://www.tdwg.org/standards/638```), we could probably get away with some pattern like ```http://rs.tdwg.org/ac/docs/``` as the subpath within which Audubon Core standards documents would be found.  So for example, the [Audubon Core structure document](https://terms.tdwg.org/wiki/Audubon_Core_Structure) could be represented by a URI like ```http://rs.tdwg.org/ac/guides/structure/```.  Under that model, I'd recommend using ```http://rs.tdwg.org/dwc/guides/text/``` for the text guide and ```http://rs.tdwg.org/dwc/guides/changes/``` for the modified change policy.  If desired, one could have a different subpath for documents that are associated with the standard, but not part of it, such as ```http://rs.tdwg.org/dwc/docs/quick/``` for the quick reference guide and ```http://rs.tdwg.org/dwc/docs/ancillary/``` for the ancillary documents landing page.  

Note that adopting these URI patterns for the documents as abstract entities does not require their representations to be in any particular web location or format.  Following patterns like this will make it easier to automatically generate the machine-readable metadata by a script, while the actual human-readable document could be anywhere (mediawiki, a GitHub readme) and in any format (markdown or html) because the server would redirect from the abstract URIs to the concrete representations when content-type: text/html is requested by a browser.

Thoughts?

**Comment by peterdesmet** [comment]
@baskaufs @mdoering the redirect rules I defined at the very top are obsolete (was work from before the October hackathon). I just reused this open issue because we'll still to do some redirecting.

@baskaufs I trust your judgement on URI patterns and we should stick to those. Idem @mdoering for not having `index.html` in the URL.

The specific issue here is: we need to host the quick reference guide somewhere (and only that document: all the rest can be served from the GitHub repository). The easiest way to host is from the [gh-pages branch](https://github.com/tdwg/dwc/tree/gh-pages), at https://tdwg.github.io/dwc/www/guides/ That branch however currently hosts everything at http://rs.tdwg.org/dwc/ If we update it from master a lot links will break.

1. We therefor need to define which current working links we want to preserve, i.e. redirect... and where to. I've shared a Google Spreadsheet with you to define those rules before implementing them
2. Once set up, we can push master to gh-pages, which will serve https://tdwg.github.io/dwc/www/guides/ Since we only have one page for the Darwin Core website, it might actually be better if it's stored in the root and served from https://tdwg.github.io/dwc/ What do you think?

**Comment by baskaufs** [comment]
@peterdesmet @mdoering So here was what I was envisioning would eventually happen.  All human-readable, human-curated pages would be written in Markdown and kept in some logical place in the TDWG Github repo.  All of the dereferencing of anything in the http://rs.tdwg.org/ subdomain would be handled by the BaseX server that generates the machine-readable metadata (which @mdoering probably doesn't know anything about at this point).  In the case of the previously mentioned human-curated pages, there would be a redirect to the markdown pages (probably a 303 redirect, but I'd have to test browser behavior to make sure).  Many of the human-readable pages that the Standards Documentation Spec say should exist (like lists of what terms are on term lists) would be minimal HTML pages generated automatically by a script in a manner similar to how the machine-readable metadata are delivered.  

Obviously this will take some work to implement and I can chip away at working on it once the BaseX server is set up.  So the real issue is how to move to using the new quick reference guide while still keeping the old HTML pages accessible.  I think putting it at tdwg.github.io/dwc/ is probably OK.

It seems to me that the patterns you have put in that Google sheet should be fine.  Having the existing URLs in the first column redirect to your suggested ones should prevent people's links from breaking.  What would come into play later would be to get the URLs listed correctly in the document.  That is, the document would list the "permanent" URLs that redirect to the Markdown pages, wherever they are.  People should bookmark those permanent URLs rather than the actual locations of the Markdown pages since those Markdown page URLs might change in the future (or maybe not even be in Markdown when something different comes along).  Of course, one wouldn't want people to start bookmarking them until they actually worked, which would be a reason to get the BaseX server operating as soon as possible.

**Comment by peterdesmet** [comment]
Ok, updated 4 first links in spreadsheet.

> So the real issue is how to move to using the new quick reference guide while still keeping the old HTML pages accessible.

That is not possible: any commit to gh-pages would remove everything.

**Comment by baskaufs** [comment]
Hmm.  How about something like this:

1. Use Pandoc or something similar to transform the four or so HTML pages that we care about into Markdown.  
2. Set up the BaseX server, and for the time being just create the script necessary to redirect to the pages we care about.
3. Try flipping the switch to redirect all traffic from rs.tdwg.org through it.  That would actually be a useful activity because if the BaseX server isn't robust enough to handle that traffic, there is no point in trying to write the more complicated scripts to do all of the other stuff.  If it fails, we fall back on what we have now and create a plan B.  

For the benefit of @mdoering, what needs to happen is to have the BaseX jar file from http://basex.org/products/download/all-downloads/ installed on some server.  I am not sure about what needs to happen with respect to a front-end server.  At Vanderbilt, we have an Nginx server (in front of a Tomcat server that's running BaseX) taking care of the secure HTTP.  However, that brings up the issue that came up before about how many things would get broken if all URIs got changed from http:// to https:// .  I'm thinking that this might be a solution:

1. Have the front end server use https://rs.tdwg.org/ and pass the requests through to the Tomcat (or whatever) server that is running BaseX.
2. Have GBIF server (or whatever is currently taking care of the tdwg.org domain) do a 301 [moved permanently) redirect from any http://rs.tdwg.org/ request to https://rs.tdwg.org/ (the front end of the BaseX server).  

After that is implemented, it wouldn't really matter if people started bookmarking the https:// versions of URIs as long as the old http:// ones always redirected.  People could and should continue to use the http:// URIs as persistent identifiers, and NOT use the https:// ones as identifiers.

I hope this makes sense.  I am near my limits on understanding the intricacies of server behavior.

**Comment by mdoering** [comment]
thanks @baskaufs, looks like I need to read up a lot more to understand all the changes you guys have been working on so actively in the past weeks. Where can I best catch up with the plan to use basex for TDWG? Should this run on a GBIF server or somewhere else? If GBIF, do you have any idea on what resources (CPU, Mem, Storage) are needed for it? I see it is also available as a [docker image](https://github.com/BaseXdb/basex#docker-image)

This question might be premature, but what is the status of https://github.com/tdwg/infrastructure/tree/master/vhosts once https://github.com/tdwg/rs.tdwg.org is to go live, is it a replacement?

**Comment by baskaufs** [comment]
So, here's the situation.  I've written scripts necessary to perform content negotiation and provide machine-readable metadata when it's requested (RDF/XML, RDF/Turtle, and JSON-LD).  I did it using BaseX for reasons I explained at http://baskauf.blogspot.com/2017/03/a-web-service-with-content-negotiation.html .  The server script that makes this happen can also do redirects and it's possible to generate HTML by script from the same XML database that provides the source metadata for the machine-readable representations.

This is not the only possible way to do it, but if we don't use the system I developed using BaseX, then somebody else will have to do the coding and that's probably not going to happen anytime in the foreseeable future.  I have the system running at Vanderbilt on a cloud server, so I know that it works, but there isn't a lot of traffic, so I don't know how robust it is.  Whether it is best to operate it from a GBIF server or elsewhere is not something I know - I was assuming a GBIF server because that's where hosting for other TDWG stuff was happening.  But from my point of view, it doesn't really matter as long as I can load the necessary data and access the configuration file that controls the dereferencing behavior.   

There is a description of how we set up our cloud server at http://baskauf.blogspot.com/2017/07/how-and-why-we-set-up-sparql-endpoint.html .  That blog post is about the SPARQL endpoint but we also have a BaseX server running as a different context of the same Tomcat  instance.  The screenshots show the reserved and scaling limits we are using.  There have only been occasional times when those resources weren't enough and those times were when somebody ran a query that got hung up - that wouldn't be an issue here because we aren't talking about setting up a Blazegraph endpoint.  But there would probably be more traffic for TDWG, particularly people wanting to look at the DwC reference documents.

The storage requirements would be rather low.  The actual data that has to be loaded for it to work only takes up 2.2 Mb as CSV files.  It's actually loaded as XML, which might be a bit bigger with the tags, but that should give you an idea of the scale.  Most of the storage space in our installation is taken up by the server software - the data are pretty minimal.  

The way we manage loading data is through HTTP POST, which is blocked without secure authentication.      However, I access the script that makes the decisions about how to dereference various URI patterns by actually logging onto the Servint control panel to edit the file directly.  There might be a better way if the server was configured elsewhere.  We didn't use Docker because the setup we used was simpler to set up and maintain, but I'm sure it would be a possibility.

**Comment by MattBlissett** [comment]
> http://rs.tdwg.org/dwc/terms/index.htm#basisOfRecord | dwc.tdwg.org/terms/#dwc:basisOfRecord (note: dwc:)

This isn't possible, the server isn't sent the anchor part (`#basisOfRecord`) so it can't change it.

**Comment by hlapp** [comment]
BTW I know it's kind of late chiming in on the basics of this, but it seems like a very bad idea for TDWG to me (and one that TDWG has gone down in the past with very poor long-term outcomes) to stand up server infrastructure that uses technologies rarely or not at all used in its constituent community (such as BaseX), and that therefore have a single point of failure (the one person who happens to understand how to maintain it). BaseX may be a fine choice for Vanderbilt given their Digital Humanities and related activities, but it seems like a perilous choice to make for TDWG.

**Comment by baskaufs** [comment]
I totally get what you are saying, but if we don't use what I'm creating, how is implementing content negotiation for the TDWG resources going to happen in the near future?  I have literally donated hundreds of hours of free time to making this work, and if it doesn't get used, then somebody else is going to have to donate their time, or somebody is going to have to fund paying somebody to do it.  It's the system we have.

The point of the design is that all of the source data comes from [CSVs on Github](https://github.com/tdwg/rs.tdwg.org).  That means those data are platform-independent.  Somebody can and should create a better delivery system, and the work done to assemble the metadata won't have been done in vain.

So if you or anybody else want to create a better system, go for it, and let it replace what I've made when its finished.  XQuery is, by the way, a W3C standard, unlike other technologies we've used (LSID for example).

**Comment by hlapp** [comment]
I guess my point is that having rs.tdwg.org provide content negotiation in a way that's not long-term sustainable is _not better_ than not having it, because I can't build anything on top of it (and that therefore relies on it) anyway. It's not abut whether XQuery is a good standard, or whether BaseX is a nice tool or not. 

TDWG once had the Twiki wiki. Our IG used it and put a lot of time into it, with disastrous results. (The Twiki is gone now because once the original contractor who set it up was gone nobody at TDWG had the knowledge of maintaining it.) This was and in fact continues to be extraordinarily painful.

I think if anyone wants to build something that they may want TDWG to adopt or take over in the future, it _must_ be done in a way that TDWG can do so sustainably. I.e., for whoever goes off and starts such a project, which technologies happen to be favorable at their home institutions for whatever reasons doesn't matter. The technology need to be sustainably adoptable by TDWG. And if you worry about that only after the fact rather than from the outset, then I'm afraid that shouldn't be made TDWG's problem.

**Comment by baskaufs** [comment]
Well, Hilmar, next time, get involved in the process.  It hasn't been a secret what I've been working on and I've talked to the relevant stakeholders in DwC, Audubon Core, the TAG, and GBIF and have gotten the green light from all of them.  Don't wait until the day before a project is done to start giving feedback.

**Comment by hlapp** [comment]
Sorry @baskaufs if I missed the fact that the feedback period had been closed. Also sorry that I missed the fact this has been talked about with relevant stakeholders (I will say though that it certainly hasn't been talked about among the TAG for the past 12 months, or least not in a way that involved the TAG membership). It seemed you were asking for feedback, the thread seemed to be open for feedback, so I gave it.

That's all, and I have to say I don't understand how I share blame for you or this project being in the place that it's in. Technical sustainability isn't exactly a radical new concept, and the TAG has made a particular point of it for any revamp of the technology stack powering the website and thereby the "wiki". Why would rs.tdwg.org be exempt from this consideration?

**Comment by peterdesmet** [comment]
@tucotuco @baskaufs all redirects listed above are functional, except for  http://rs.tdwg.org/dwc/terms/history/decisions/index.htm: should I close this issue and discuss what to do with decisions in a separate issue?

**Comment by tucotuco** [comment]
Yes.

**Comment by baskaufs** [comment]
I think decisions should be a separate issue.  When the decisions list started, Darwin Core was the only vocabulary where decisions were required.  However, now there is Audubon Core and probably other controlled vocabularies in the future, and we probably don't want to maintain separate decisions lists for them.  I was envisioning using http://rs.tdwg.org/decisions/decision-2009-12-07_1 as the pattern (local name the same, but outside of the DwC namespace).

**Comment by baskaufs** [comment]
This issue is old enough that I'm not exactly sure which redirects @peterdesmet is talking about.

**Comment by tucotuco** [comment]
I agree with the sentiment and the pattern.
